spmd/tensor/device_mesh.py

```diff
@@ -69,61 +95,114 @@ def __init__(self, device_type, mesh):
                 f"DeviceMesh only support cpu or cuda device type, but got {device_type}"
             )
 
-        # TODO: support multi-dimensional device mesh
-        assert self.mesh.ndim == 1, "Only support 1-d device mesh for now"
+        world_size = get_world_size()
+        if self.mesh.numel() > world_size:
+            raise RuntimeError(
+                f"Mesh should not be bigger than default world size, but found {self.mesh.numel()} ranks!"
+            )
 
-    # pyre-fixme[3]: Return type must be annotated.
-    def __enter__(self):
+        unique_mesh_values = self.mesh.unique(sorted=True)
+        if unique_mesh_values.numel() != self.mesh.numel():
+            raise RuntimeError(
+                f"DeviceMesh cannot have duplicate values, but found {self.mesh.tolist()}"
+            )
+
+        if dim_groups is not None:
+            # if user hand creating dimension based groups
+            # we just take it and use it for communication
+            # we assume user passing dim_gruops are legit
+            # TODO: add more checks to check the correctness
+            # of user passed in dim groups
+            self._dim_to_groups = dim_groups
+            return
+
+        self._dim_to_groups: List[List[ProcessGroup]] = [
+            [] for _ in range(self.mesh.ndim)
+        ]
+        # create sub pgs base on the mesh argument
+        for dim in range(self.mesh.ndim):
+            subgroups_by_dim = self._dim_to_groups[dim]
+            # generate flattened mesh for dim, to create subgroups
+            unbinded_mesh = self.mesh.unbind(dim)
+            flattened_mesh = [submesh.flatten() for submesh in unbinded_mesh]
+            stacked_mesh = torch.stack(flattened_mesh)
+
+            if self.mesh.ndim == 1:
+                if unique_mesh_values[-1] == world_size - 1:
+                    # we just append the default pg to the first
+                    # dim goups, as new_group cannot have the
+                    # exact same ranks as world
+                    subgroups_by_dim.append(default_pg)
+                else:
+                    # smaller than world, create new group
+                    subgroups_by_dim.append(
+                        new_group(ranks=self.mesh, backend=backend_name)
+                    )
+            else:
+                # multi-dim mesh, create subgroups by
+                # looping over the transposed stacked_mesh
+                # and append the groups
+                for dim_mesh in stacked_mesh.T:
```

Comment:
Why is T needed here?

For your test example, if the mesh is like:
torch.arange(8).reshape(2, 2, 2)
tensor([[[0, 1],
         [2, 3]],
        [[4, 5],
         [6, 7]]])
So stacked_mesh will be
tensor([[0, 1, 2, 3],
        [4, 5, 6, 7]])

And stacked_mesh.T becomes:
tensor([[0, 4],
        [1, 5],
        [2, 6],
        [3, 7]])

Any reason why we use T not the original tensor to create the sub-process group? I personally think using the original tensor makes more sense. But this is not a blocker just out of curiosity.

Reply:
The reason we have this transpose is that this works for 1d and 2d to generate corresponding submeshes, and I am relying on induction rules to ensure n-d works. 

For 3d, it's a bit hard to visualize, we can think of 3d as a cube that have the 3rd dimension, given the 3rd dim size = 2, it should be something like 

([[0, 4],
[1, 5],
[2, 6],
[3, 7]])

If we only look over through the 3rd dimension. 

For more than 3d, it's hard and I am purely relying on induction.

Reply:
Would the following work? 
ranks_per_mesh = mesh.swapdims(-1, dim).reshape(-1, mesh.shape(dim))
for dim_mesh in ranks_per_mesh:
    new_group(..)

swapdims would place the current dim as the inner-most dimension of the tensor, and then reshape would flatten all other dimensions out.

Also, I think that new_group will only return a valid process group for one of the calls in this loop, since each rank will belong to only one process group per mesh dimension, so maybe we can check for that and make subgroups_by_dim a List[PG] instead of List[List[PG]] (maybe call them local_subgroup_by_dim)

Reply:
Yeah that looks better, especially the single pg per dim could help us simplify things, will update, thanks for suggestion!